Fix openapi schema xml comments handling for referenced schemas

desjoerd opened the pull request:
# Fix openapi schema xml comments handling for nested schemas

Fix applying annotations from xml on properties which would apply to referenced schemas.

## Description

Changes:
- Change the order of applying XML comments on schemas to make sure the property comment is leading (for applying the annotation on the reference)
- Don't apply property comments on referenced schemas because this causes issues when a schema is referenced multiple times, it would get the description from the last property.

Fixes #61965

desjoerd committed:
Update OpenAPI range formatting to format using the target culture and Update tests to write in the InvariantCulture

desjoerd committed:
Updated schema reference XML Comment handling.

Update handling of nested schemas and referenced schemas

martincostello commented:
Did you mean to include 14dd22b9845c510b78bcc2bf6a296af621a4e7c5 in this PR?

desjoerd commented:
No, I am planning to rebase/cherry-pick when #62212 is merged. Otherwise I cannot run the unit tests 😅. I marked it as draft because it's also not completely correct at the moment. There needs to be some extra conditions to check whether it's a reference or not (to handle the two different descriptions, on the property with reference and on the schema in components).

I currently only get either the correct description on the reference and wrong on the schema or the other way around.

I am open to have a call about it, maybe tomorrow? Or to continue the discussion here or on the referenced issue.

desjoerd committed:
Only set a schema property description when it's not a schema reference

Update XmlCommentGenerator.Emitter.cs to skip applying property descriptions when it's a schema reference.

Schema transformers get the full schema and not the schema reference. So when the description is updated it would update it for all locations.

desjoerd committed:
Snapshot the generated openapi as part of the OpenApi source generator tests

desjoerd commented:
For now I fixed the issue of having two schema's inherit the last property comment as a description by excluding referenced schemas from getting the description from the property. This can be seen as a downgrade in functionality, but it would otherwise give wrong output.

A schema transformer doesn't get a hold of the actual SchemaReference itself but after transforming it get's resolved to a reference in the `OpenApiSchemaService.ResolveReferenceForSchema`. To support getting the description on the SchemaReference it requires a very delicate order of the schema transformers. This would be quite a big change.

The SchemaReference description should also ONLY come from the property and not from the type, as that would give duplicate descriptions and a noisy openapi specification.

desjoerd committed:
Uncomment and remove commented code

desjoerd commented on src/OpenApi/test/Microsoft.AspNetCore.OpenApi.SourceGenerators.Tests/CompletenessTests.cs:
```diff
@@ -516,9 +516,6 @@ await SnapshotTestHelper.VerifyOpenApi(compilation, document =>
             Assert.Equal("This class validates the behavior for mapping\ngeneric types to open generics for use in\ntypeof expressions.", genericParent.Description, ignoreLineEndingDifferences: true);
             Assert.Equal("This property is a nullable value type.", genericParent.Properties["id"].Description);
             Assert.Equal("This property is a nullable reference type.", genericParent.Properties["name"].Description);
-            Assert.Equal("This property is a generic type containing a tuple.", genericParent.Properties["taskOfTupleProp"].Description);
```
I had to remove these assertions as the schema is a reference and the comment is from a property.

Other solution/fix would be to inline these schemas by adding a `CreateSchemaReferenceId` which returns `null` for Tuples.

desjoerd committed:
Rename emitted variable to isInlinedSchema

desjoerd commented on src/OpenApi/test/Microsoft.AspNetCore.OpenApi.SourceGenerators.Tests/SnapshotTestHelper.cs:
```diff
@@ -200,6 +200,14 @@ void OnEntryPointExit(Exception exception)
             var targetMethod = serviceType.GetMethod("GenerateAsync", [typeof(string), typeof(TextWriter)]) ?? throw new InvalidOperationException("Could not resolve GenerateAsync method.");
             targetMethod.Invoke(service, ["v1", writer]);
+
+            var openApiString = Encoding.UTF8.GetString(stream.ToArray());
```
If required or wanted I can extract this to a seperate PR. Testing and verifying openapi documents was hard without this. The reason is, it is hard to assert when it generates too much in the openapi document.

captainsafia replied:
I'd usually welcome changes like this but after https://github.com/dotnet/aspnetcore/issues/63073 I'm a little bit cautious 😅 

Best to put this in a separate PR just to be safe.

desjoerd commented:
@captainsafia it works but it would mean creating `x-reference-<annotation>` for all annotations possible on a reference (https://json-schema.org/understanding-json-schema/reference/annotations#annotations).

I created a PR on top of this one, very quick and dirty, but for the gist: https://github.com/desjoerd/aspnetcore/pull/11

I think the issue is also present for [Attributes] in https://github.com/dotnet/aspnetcore/blob/406950442014149931322f97c42e999ce03fdb71/src/OpenApi/src/Services/Schemas/OpenApiSchemaService.cs#L100-L114 because that is also applied on all referenced schemas. When a schema is used multiple times in different properties it would override the description, but it should actually be on the reference.

The moment we add all `x-reference-<annotation>` we enable modifying the Schema Reference from schema transformers, but it's not a nice API like `OpenApiSchemaReference`.

I think the choice is between exposing the attributes of `OpenApiSchemaReference` by using Metadata vs some other way. The other way is harder to do because that would require a .NET Api change instead of a behavioral api change.

@martincostello maybe you have some ideas as well?

desjoerd commented:
I've been thinking, the options I can think of are:

1. Transform IJsonSchema instead of the actual class which is implemented by both the schema and the reference. It's binary incompatible but it should not affect new compiled code as it has the same properties.

2. Add the JsonSchemaReference to the context Schema Transformer Context.

3. Work with Metadata and hoist it over to the reference when resolving.

As a library author and customizer of my OpenApi I would prefer to be able to do almost everything from transformers or the exposed Api via the OpenApi Options. Things like applying data annotations are now internal, but I think they could be done as Schema Transformers. Giving the option also to disable DataAnnotations support for Openapi (for example when using Fluent Validations).

(ps: braindump from my phone)

captainsafia commented:
Yeah, I think in an ideal world we would expand the API for schema transformers to support application on both target schemas and schema references but that would require some considerable API work and likely wouldn't land in .NET 10 at this point. I figure that special casing handling for summary/description on XML comments might be a suitable tactical change.

I do like the idea of shifting more things out of the JsonNode layer to the OpenApiSchema -- particular the validation attributes you mentioned. I could see a model where these defaults are registered as transformers that users can add/remove from their OpenApiOptions.

With regard the the options presented, I considered #1 but wasn't sure if that would be confusing considering there is a legacy of transformers/filters/processors applying on raw schemas and not references. However, that decision does predate the existence of OpenApiSchemaReference.

#2 feels doable. I imagine the new property would be nullable. When non-null it indicates the transformer is applying on a reference. In the implementation, we would have two passes of transformers, on the target and the reference. This one seems doable to fit in for .NET 10 and might solve this problem.

(ditto: brain dump from my phone 😅)

desjoerd commented:
For Option 1: I forgot that only the `get` is exposed on [IOpenApiSchema](https://github.com/microsoft/OpenAPI.NET/blob/main/src/Microsoft.OpenApi/Models/Interfaces/IOpenApiSchema.cs). So that would be a breaking change for all existing transformers. Requiring a typecheck and cast before transforming. So I think this is unwanted.

Maybe then Option 2 is best. Only weird thing is that it is a property on the Context (which should be treated as read-only), I've seen some confusion around this. Could also be an extra parameter `(schema, schemaReference, context) => ` but that is also a breaking change, but less impactfull then Option 1. 
On the multiple passes, it's one extra as there is already a transform for each property, only not for the schema treated as a root itself. In theory for this instance it would not even need an extra pass on the schema alone.

I think we can do either one of the options as a follow up PR. And treat this PR as a mitigation for modifying the referenced schema from the property comments.

I think the follow ups would be:
- Check for Attributes as well whether the same problem is there for referenced schemas (inheriting comments & metadata from properties where the last one wins)
- Extend the api (or strings) to allow customizing references
- Enhance the openapi references based on the property metadata.

captainsafia commented:
> Only weird thing is that it is a property on the Context (which should be treated as read-only), I've seen some confusion around this. Could also be an extra parameter (schema, schemaReference, context) =>  but that is also a breaking change, but less impactfull then Option 1.

Another option is to introduce a completely new transformer type `IOpenApISchemaReferenceTransformer` with it's own context object and transform signature and avoid clobbering over the existing one.

> I think the follow ups would be:

Agree with this ordering. For 2, we might have to rely on the metadata approach to start given the need to do API review, especially with RC1 snap coming closer.

desjoerd commented:
I thought so in regards to Api changes.

I was also thinking about a Schema Reference Transformer. I do not know whether it is allowed to have an Api change so late in this stage, else it's for .NET 11.

Tomorrow evening (CEST) I've got some time and will add the schema "Metadata" approach in this PR and do some checks on the Description attribute. Then it will be correct and have the ability to set the description for references.

Only note is for the Metadata approach that people (like me 😅) should not depend on it because it will change when adding a "neat" way. But then again, adding annotations on references is something which is OpenApi 3.1 specific. In Openapi 3.0 a ref completely replaces the whole object.

captainsafia commented:
> Only note is for the Metadata approach that people (like me 😅) should not depend on it because it will change when adding a "neat" way.

Yeah, admittedly, we don't do anything very aggressive to guard against people taking a dependency on the values we populate into Metadata. They're just harder to discover because they are never serialized. We may want to put a note in the docs somewhere that populated metadata is an implementation detail and not part of the public API contract.

desjoerd committed:
Add handling of XML Comments on properties with a schema reference with Metadata properties

desjoerd committed:
Improve XmlComments SchemaReference tests and revert modified unrelated tests

desjoerd committed:
Revert outputting OpenApi.json snapshots in the XmlSourceGenerator Tests

desjoerd committed:
Fix dereference null warning because the compiler cannot analyze it from a boolean value

desjoerd committed:
Fix null assignment warning

desjoerd committed:
Add tests around <example> comments

desjoerd commented:
@captainsafia I've pushed my changes.

- Removed the openapi.json snapshots
- Add metadata properties for the reference description and example
- Added unit tests and reverted some removed assertions as eveything arounc XML comments should be working.

Remarks:
I noticed that for the `[Description]` attribute we've got the same problem around OpenApiReferences (where the last property Description wins).

Also with the current ordering of applying comments and the `[Description]` attribute would mean that the XML comments win. Where I would say it should be the `[Description]` attribute should win.

I would say, let's address these in follow ups as well.

desjoerd commented on src/OpenApi/gen/XmlCommentGenerator.Emitter.cs:
```diff
@@ -468,6 +458,34 @@ public Task TransformAsync(OpenApiSchema schema, OpenApiSchemaTransformerContext
+                        || string.IsNullOrEmpty(schemaId as string))
+                    {
+                        // Inlined schema
+                        schema.Description = propertyComment.Value ?? propertyComment.Returns ?? propertyComment.Summary!;
```
I can live with this

captainsafia replied:
I'm surprised that a null suppression operator was needed here since `schema.Description` is a nullable string. Was the null suppression required for some reason?

desjoerd commented on src/OpenApi/gen/XmlCommentGenerator.Emitter.cs:
```diff
@@ -468,6 +458,34 @@ public Task TransformAsync(OpenApiSchema schema, OpenApiSchemaTransformerContext
+                    else
+                    {
+                        // Schema Reference
+                        schema.Metadata["x-ref-description"] = propertyComment.Value ?? propertyComment.Returns ?? propertyComment.Summary!;
```
This is type asserted to a string where it's used

desjoerd commented on src/OpenApi/gen/XmlCommentGenerator.Emitter.cs:
```diff
@@ -468,6 +458,34 @@ public Task TransformAsync(OpenApiSchema schema, OpenApiSchemaTransformerContext
+                        schema.Metadata["x-ref-description"] = propertyComment.Value ?? propertyComment.Returns ?? propertyComment.Summary!;
+                        if (propertyComment.Examples?.FirstOrDefault() is { } jsonString)
+                        {
+                            schema.Metadata["x-ref-example"] = jsonString.Parse()!;
```
Dito

desjoerd commented on src/OpenApi/gen/XmlCommentGenerator.Emitter.cs:
```diff
@@ -468,6 +458,34 @@ public Task TransformAsync(OpenApiSchema schema, OpenApiSchemaTransformerContext
+                // Apply comments from the property
+                if (XmlCommentCache.Cache.TryGetValue(DocumentationCommentIdHelper.NormalizeDocId(propertyInfo.CreateDocumentationId()), out var propertyComment))
+                {
+                    if(schema.Metadata is null
```
It's doing that.

desjoerd committed:
Fix formatting of emitted code

desjoerd commented on src/OpenApi/gen/XmlCommentGenerator.Emitter.cs:
```diff
@@ -468,6 +458,34 @@ public Task TransformAsync(OpenApiSchema schema, OpenApiSchemaTransformerContext
+                // Apply comments from the property
+                if (XmlCommentCache.Cache.TryGetValue(DocumentationCommentIdHelper.NormalizeDocId(propertyInfo.CreateDocumentationId()), out var propertyComment))
+                {
+                    if(schema.Metadata is null
```
Fixed

captainsafia commented on src/OpenApi/src/Services/Schemas/OpenApiSchemaService.cs:
```diff
@@ -101,7 +101,8 @@ internal sealed class OpenApiSchemaService(
             }
-            if (context.PropertyInfo is { AttributeProvider: { } attributeProvider })
+            var isInlinedSchema = schema["x-schema-id"] is null;
+            if (isInlinedSchema && context.PropertyInfo is { AttributeProvider: { } attributeProvider })
```
Why is `isInlinedSchema` added here? I'm slightly surprised this doesn't break any tests since this would mean that we wouldn't apply attributes to properties that are referenced by ref. A scenario like the following:

```csharp
public class ComplexType
{
  [Description("FooBarBaz")]
  public AnotherComplexType PropertyName { get; }
}
```

desjoerd replied:
I was suprised as well that the Unit Tests where still passing. I put this in by accident for the follow-up PR to fix the Description annotations.

They are also applied to the Actual Schemas (the last property description wins).

I will revert this to keep it focused.

desjoerd committed:
Revert change to OpenApiSchemaService

desjoerd commented:
@captainsafia I've reverted the change on the schema service, will be part of a follow-up 😎 (Something like: https://github.com/desjoerd/aspnetcore/pull/12).
